src/librustc_codegen_llvm/intrinsic.rs

```diff
@@ -695,6 +696,23 @@ impl IntrinsicCallMethods<'tcx> for Builder<'a, 'll, 'tcx> {
                 return;
             }
 
+            "freeze" => {
+                let dst = args[0].immediate();
+                let r = self.inline_asm_call(
+                    CStr::from_bytes_with_nul(b"\0").unwrap(),
+                    CStr::from_bytes_with_nul(b"r,~{memory}\0").unwrap(),
+                    &[dst],
+                    self.type_void(),
+                    true,
+                    false,
+                    AsmDialect::Att,
+                );
```

Comment:
We could optimize this when `count` is a constant to avoid the memory clobber by casting `dst` to a `[u8; count]` and using `asm!("" : "=*m" (dst) : "0" (dst))`.

A memory clobber can be expensive since it forces the compiler to reload the values of all values that are potentially globally visible.

This optimization requires `count` to be properly const-propagated through inlining. Can we do this?

Reply:
There are inlining and (limited) constant propagation passes for MIR but the former is not enabled by default and it's not clear when it will be.

I can't find indiction that your proposal will actually result in improved codegen (see [this test case](https://godbolt.org/z/H-fsWh)). Pointee types in LLVM IR are meaningless and slated for removal (if only someone would push that refactoring over the finish line), so I see no reason why the LLVM IR pointee type should have bearing on how many bytes the inline asm can write through the pointer.

Now, if there was a way to indicate that the inline asm guarantees it only accesses a certain range of bytes around the address that's passed in, that would help, but AFAIK there's no such thing.

Reply:
Hmm, I tried this out in C: https://godbolt.org/z/Fbz1o5

It seems that GCC interprets "m" constraints very precisely, while it seems that LLVM just treats all memory constraints (even read-only memory inputs) as a general memory clobber.

I suppose that in this case we can just stick to the existing formulation since there is no advantage in optimizing it.

Reply:
Yeah, it seems like this is just going to be inherently a bit imprecise until LLVM lands freeze support on its end.

Reply:
I cannot review this part, as I have no knowledge of these assembly annotation.

@Amanieu could you confirm that this tells the compiler that the assembly block may read and/or write all memory reachable from `dst`, such that it can neither optimize away preceding writes nor following loads?

Reply:
> could you confirm that this tells the compiler that the assembly block may read and/or write all memory reachable from `dst`, such that it can neither optimize away preceding writes nor following loads?

@comex put it succinctly here: https://github.com/rust-lang/rfcs/pull/2360#issuecomment-466607141

`memory` makes LLVM treats this as `readnone = readonly = false`, which means that this can read and write to all memory that's not private. This includes all globals, heap memory, etc. and AFAICT definitely includes all memory reachable from `dst`.